### PR TITLE
TRT-2814: Eliminate prow_job_runs_report_matview

### PR DESCRIPTION
**Pull request by mstaeble**
## Summary

- Replaces `prow_job_runs_report_matview` with a two-phase direct query in `JobsRunsReportFromDB`: Phase 1 paginates from base tables (`prow_job_runs` + `prow_jobs`), Phase 2 enriches the paginated page with test name arrays, pull request data, and annotations via `errgroup`.
- Removes the matview definition and its entry from `PostgresMatViews`, eliminating a ~90-second refresh that blocked data loading.
- Reads `test_failures` and `test_flakes` directly from stored columns on `prow_job_runs` (populated at insert time by the prow loader), eliminating the runtime subqueries the matview used to compute these counts.
- Refactors `pkg/filter` to export `FilterItemToSQL` (column-expression based) and `FilterFieldToSQL` (type-aware with array/timestamp handling), replacing the unexported `orFilterToSQL`.
- Adds `sortable: false` to test name columns in the frontend and returns `ValidationError` for unsortable sort fields.
- Fixes ILIKE wildcard escaping globally: consolidates `escapeLikeMetachars` into an exported `filter.EscapeLikeMetachars` and applies it at all 12 ILIKE pattern construction sites (`FilterItemToSQL`, `FilterFieldToSQL`, `andFilterToSQL`, `testNameFilterSQL`). Previously, user-provided `%` and `_` characters in filter values acted as ILIKE wildcards, causing unintended matches.

## Key design decisions

- **Stored counts, not runtime aggregation**: `test_failures` and `test_flakes` are read directly from `prow_job_runs` columns rather than computed via subqueries or CTEs. These columns are populated at insert time by the prow loader ([TRT-2834](https://redhat.atlassian.net/browse/TRT-2834)), so no runtime aggregation is needed.
- **Conditional JOINs**: PR join is added only when needed for filtering or sorting, split into filter-required (before COUNT) and sort-only (after COUNT) to avoid unnecessary work during pagination counting.
- **NULL-safe HasEntry**: Uses `COALESCE(col, '{}')` to ensure `NOT` wrapping works correctly for NULL arrays.
- **Test name filtering**: Uses EXISTS subqueries against `prow_job_run_tests` rather than array column filters, matching the previous matview behavior.

## Test plan

- [x] Lint passes (`make lint`)
- [x] Unit tests pass (`make test`)
- [x] Integration tests pass (`make integration`) with 43 test functions covering:
  - Basic query, pagination, release filtering, timestamp windowing
  - Sorting by timestamp, test_failures, test_flakes, job name
  - All filter operators: `=`, `!=`, `>`, `>=`, `<`, `<=`, `contains`, `startsWith`, `endsWith`, `hasEntry`, `hasEntryContaining`, `isEmpty`, `isNotEmpty`
  - All filter dispatch paths: test name EXISTS, column aliases, generic filter
  - Fields: job, brief_name, cluster, overall_result, variants, timestamp (epoch), test_failures, test_flakes, failed/flaked/ran_test_names, pull_request_* fields
  - AND/OR link operators, NOT negation, cross-dispatch-path OR combinations
  - Pagination edge cases (beyond results, partial last page)
  - Enrichment: annotations, pull requests, test name arrays
  - Boundary conditions: report end exclusion, lookback window
  - Validation: unsortable field rejection
- [x] Independent code review found no correctness issues
- [x] Independent test coverage analysis verified all filter paths and operators are tested
- [x] Verified against staging database (20,446 runs for release 4.19)

## Staging verification (sippy-staging, release 4.19)

All 10 checks passed against the staging PostgreSQL database with real production-scale data:

| # | Test | Result |
|---|------|--------|
| 1 | Basic paginated query (perPage=5, sort=timestamp desc) | 20,446 total rows, correct fields (brief_name, variants, cluster, timestamp) |
| 2 | Pagination consistency (page 0 vs page 1) | Stable TotalRows, zero overlap between pages |
| 3 | Job filter (`job contains "aws"`) | 6,307 results, all job names contain "aws" |
| 4 | NOT negation (`NOT job contains "aws"`) | 14,139 results, no job names contain "aws" |
| 5 | Test name filter + enrichment (`failed_test_names contains "etcd"`) | 177 results, failed_test_names populated with matching etcd tests |
| 6 | ILIKE escaping (`e2e_aws` vs `e2e-aws`) | Literal underscore: 0 results. Hyphen: 5,164 results. Confirms escaping works. |
| 7 | Annotations enrichment | 6/25 runs have annotations (e.g. `release.openshift.io/architecture`) |
| 8 | Sort by test_failures descending | Correctly ordered: 650, 361, 324, 314, 302 |
| 9 | Zero matching rows (`job = "nonexistent-job-xyz"`) | TotalRows=0, empty rows, no error |
| 10 | Sort by PR field doesn't inflate TotalRows | Identical count (20,446) for sort-by-id and sort-by-pull_request_author |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Job run reports now support richer filtering and sorting, including test names, pull request fields, variants, timestamps, failures, and flakes.
  * Report results include relevant pull request metadata, annotations, and failed or flaked test names.
  * Negated filters, combined conditions, special-character searches, and pagination are handled more consistently.
  * Hidden test-name columns in the job runs table are no longer sortable.

* **Bug Fixes**
  * Prevented duplicate job runs when multiple pull requests are linked.
  * Improved lookback boundary handling and filter-value matching.

* **Tests**
  * Added comprehensive coverage for report filtering, sorting, pagination, enrichment, and edge cases.

### CODE REVIEW

**Comment by mstaeble** [comment]
@coderabbitai full review

File: pkg/api/job_runs.go [hunk 1]
```diff
@@ -177,6 +183,310 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
...
+			}
+		}
+	}
+	if len(clauses) > 0 {
+		joiner := " AND "
+		if filterOpts.Filter.LinkOperator == filter.LinkOperatorOr {
+			joiner = " OR "
+		}
+		q = q.Where(strings.Join(clauses, joiner), allArgs...)
+	}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go:278]: Fixed. Wrapped the joined clauses in explicit parentheses.

File: pkg/api/job_runs.go [hunk 2]
```diff
@@ -177,6 +183,310 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
...
+		args = []any{lookback, fmt.Sprintf("%s%%", item.Value)}
+	case filter.OperatorEndsWith:
+		sql = existsBase + " AND tests.name ILIKE ?)"
+		args = []any{lookback, fmt.Sprintf("%%%s", item.Value)}
+	default:
+		sql = existsBase + " AND tests.name ILIKE ?)"
+		args = []any{lookback, fmt.Sprintf("%%%s%%", item.Value)}
+	}
+	return filter.WrapNot(sql, item.Not), args
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go]: Fixed `testNameFilterSQL`: replaced the catch-all default with explicit `OperatorContains`/`OperatorHasEntryContaining` cases; unknown operators now return empty SQL (which `applyJobRunFilters` already skips).

The `FilterItemToSQL` `UnknownFilterOperator()` pattern is pre-existing and used across all filter paths, not introduced by this PR. Changing its signature to return an error would touch many call sites — better as a separate cleanup.

File: pkg/api/job_runs.go [hunk 3]
```diff
@@ -177,6 +183,310 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
...
+		nameMap[nameResults[i].ProwJobRunID] = &nameResults[i]
+	}
+	for i := range results {
+		if names, ok := nameMap[results[i].ID]; ok {
+			results[i].FailedTestNames = names.FailedTestNames
+			results[i].FlakedTestNames = names.FlakedTestNames
+		}
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go:359]: Not a regression. The old matview also never populated `ran_test_names` as an array — it only populated `failed_test_names` and `flaked_test_names`. The `ran_test_names` field is filter-only (via EXISTS subquery in `testNameFilterSQL`), not returned in the response. The frontend uses it for filtering ("show runs that executed test X"), not display.

File: sippy-ng/src/jobs/JobRunsTable.jsx [hunk 4]
```diff
@@ -362,20 +362,23 @@ export default function JobRunsTable(props) {
...
       hide: true,
+      sortable: false,
     },
     {
       field: 'ran_test_names',
       type: 'array',
       autocomplete: 'tests',
       headerName: 'Tests ran',
       hide: true,
+      sortable: false,
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply sippy-ng/src/jobs/JobRunsTable.jsx:381]: Sorting by test name arrays was never meaningful. The old matview sorted by the array's text representation, which produced nonsensical ordering. Any existing bookmarks using these sort fields were already broken in practice. The backend now returns a clear `ValidationError` instead of silently returning garbage order. A frontend fix to sanitize stale `sortField` params could be done as a follow-up but isn't blocking.

File: pkg/api/job_runs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go]: Good catch. The `continue` does bypass the `if sql != ""` guard.

Fixed: `testNameFilterSQL` now returns an error for unsupported operators, and `applyJobRunFilters` propagates it as a `ValidationError` before appending to clauses.

**Comment by mstaeble** [comment]
@coderabbitai resolve

**Comment by mstaeble** [comment]
@coderabbitai full review

File: pkg/api/job_runs.go [hunk 5]
```diff
@@ -45,87 +45,61 @@ const (
...
-	// Get the row count before pagination
 	var rowCount int64
 	if err := q.Count(&rowCount).Error; err != nil {
 		return nil, err
 	}
 
-	// Paginate the results:
+	if needs.prJoinForSort && !needs.prJoinForFilter {
+		q = addPRJoin(q)
+	}
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go:101]: Fixed. Added `q.Order("prow_job_runs.id DESC")` as a tiebreaker after the sort-only PR join. Also added two integration tests that exercise this directly: `TestJobRunsReport_PaginationStableWithTiedSortKey` (sorts by test_flakes where two runs tie at 0, paginates one-per-page, verifies all runs appear exactly once) and `TestJobRunsReport_PaginationDeterministicWithoutSort` (no sort field, same one-per-page verification).

File: pkg/api/job_runs.go [hunk 6]
```diff
@@ -177,6 +151,277 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
...
+			jrpp.prow_job_run_id AS id,
+			pp.link AS pull_request_link,
+			pp.sha AS pull_request_sha,
+			pp.org AS pull_request_org,
+			pp.author AS pull_request_author,
+			pp.repo AS pull_request_repo
+		FROM prow_job_run_prow_pull_requests jrpp
+			INNER JOIN prow_pull_requests pp ON pp.id = jrpp.prow_pull_request_id
+		WHERE jrpp.prow_job_run_id IN ?
+		ORDER BY jrpp.prow_job_run_id`, ids).Scan(&prResults).Error; err != nil {
```
[the remark this answers is not in the document]
Comment (mstaeble) [reply pkg/api/job_runs.go]: Fixed. Added `prow_pull_request_id DESC` as a secondary `ORDER BY` in both the `LEFT JOIN` subquery (line 75) and the enrichment `Raw` query (line 379), so both paths deterministically select the same PR for a given run.

**Comment by mstaeble** [comment]
@coderabbitai resolve

File: pkg/api/job_runs.go [hunk 7]
```diff
@@ -177,6 +153,278 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
...
+}
+
+// columnAliases maps filter field names that are SELECT aliases (not base
+// table columns) to their table-qualified column expressions. PostgreSQL
+// WHERE clauses cannot reference SELECT aliases, so these fields must be
+// rewritten before they reach the generic filter system.
+var columnAliases = map[string]string{
+	"id":                  "prow_job_runs.id",
+	"job":                 "prow_jobs.name",
+	"brief_name":          "regexp_replace(prow_jobs.name, 'periodic-ci-openshift-(multiarch|release)-master-(ci|nightly)-[0-9]+.[0-9]+-', '')",
```
Comment (neisw) [inline pkg/api/job_runs.go:204]: This is interesting as the `master` branch was renamed to `main` about 5 or 6 months ago.  So while correct in this is what the existing code has it should likely be `-(master|main)-`

Comment (neisw) [reply pkg/api/job_runs.go:204]: Presume we also have a gap where with our `synthetic` [releases](https://github.com/mstaeble/sippy/blob/7cdc7cb15292280a1f1d41e229583dea85226b7f/config/openshift-customizations.yaml).   Any idea how important the `brief_name` is?  I presume not very.

Comment (mstaeble) [reply pkg/api/job_runs.go:204]: Good eye. The `master` part of the regex should be `(master|main)` now that the repos renamed their default branch.

I traced the impact. The `brief_name` is purely cosmetic. It is used as display text in compact table views. When the regex doesn't match, the full job name passes through unchanged. There is no impact to links, routing, filtering logic, or data integrity.

The same regex appears in 3 places (twice in job_runs.go, once in the job_results SQL function in functions.go), all with the same cosmetic-only impact.

I'd prefer to defer this to a follow-up to keep this PR scoped to the matview elimination. I'll file a ticket to fix the regex globally.

By the way, synthetic release jobs (ARO HCP, ROSA, layered products, etc.) have completely different naming patterns that wouldn't match the regex even with main added. For example, periodic-ci-Azure-ARO-HCP-main-periodic-integration-e2e-parallel has no version segment to anchor on, and periodic-ci-3scale-qe-3scale-deploy-ocp4.16-... embeds the version differently. Covering those would require a broader effort beyond a regex tweak.

File: pkg/api/job_runs.go [hunk 8]
```diff
@@ -177,6 +153,278 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
 	}, nil
 }
 
+const jobRunsBaseSelect = `prow_job_runs.id,
+	prow_jobs.release,
+	prow_jobs.name,
+	prow_jobs.name AS job,
+	prow_jobs.variants,
+	regexp_replace(prow_jobs.name, 'periodic-ci-openshift-(multiarch|release)-master-(ci|nightly)-[0-9]+.[0-9]+-', '') AS brief_name,
```
Comment (neisw) [inline pkg/api/job_runs.go:161]: Just noting the `-(master|main)-` main is in multiple places

Comment (neisw) [reply pkg/api/job_runs.go:161]: Ok defering

**Comment by neisw** [comment]
/lgtm
/hold
release when ready
